### ISSUE TITLE
Metadata for services: keyword encoding guidance example for INSPIRE Part D.4 uses character string when INSPIRE recommend gmx:anchor

### ISSUE THREAD

**Issue by Sgaff** [agiorguk/gemini#29]
Hi,

In our Keyword element, for service metadata, we have an example encoding guidance of how to encode spatial data service category. Our current example however, encodes the spatial data service category keyword as a character string, whereas the INSPIRE 2017 TG recommends that the the keyword be encoded as a gmx:anchor referencing the INSPIRE Spatial Data Service Category codelist e.g <gmx:Anchor xlink:href="http://inspire.ec.europa.eu/metadatacodelist/SpatialDataServiceCategory/humanCatalogueViewer">humanCatalogViewer</gmx:Anchor>

This would be a minor change to the encoding guidance and would not involve any other textual changes, but would bring us directly in line with INSPIRE recommended best practice.

Therefore, I propose a small edit to our encoding guidance to reflect this. 

Thanks

Sean

**Comment by PeterParslow** [comment]
@Sgaff could you label this? That should help us as we begin to resolve issues

**Comment by PeterParslow** [comment]
I agree that we should make this change although it is only an INSPIRE Recommendation (not requirement) _TG Recommendation C.8: metadata/2.0/rec/common/use-anchors-for-cv-keywords_.

This affects examples 2, 3, 4 and 5 (the recommendation doesn't only apply to Spatial data service category)

As our guidelines item 3 says _Where keywords do originate from a controlled vocabulary the encoding shown in Example Two shall be used_, I would see it as breaking: an XML instance that is valid to our encoding guidelines now would not be valid after we make this change.

**Comment by PeterParslow** [comment]
Primarily Example two, but also Examples three & four.

Also separate the two sentences in item 1 of the Encoding Guidance, and bring item 3 up, to give:

1. The GEMINI2 keyword item comprises keyword value(s) and, conditionally, the specification of an originating controlled vocabulary.
2. Where keywords do originate from a controlled vocabulary the encoding shown in Example Two shall be used.
3. If keywords are not selected from a controlled vocabulary the encoding shown in Example One shall be used.
etc...

We consider this "non breaking" because the "shall" cannot be validated, being conditional on whether the term is in a controlled vocab. Perhaps drop from "shall" to "should". INSPIRE D4 keywords: it's a recommendation.

**Comment by PeterParslow** [comment]
The example "numbers" weren't appearing in the HTML output; tried to fix that - awaiting deployment to DEV to see if it worked (due to ifdef statements?)
Fixed snippets for examples 2 & 3 (there aren't examples 4 & 5 as far as I can see)
Tweaked the "encoding guidelines" wording as discussed above. Also fixed the statement about encoding when more than one keyword is from the same vocab.

Removed reference to the "incorrect" example five, which we presumably removed already.

BUT actually, I can't see how the "include" statements in the asciidoc file relate to the files they claim to include & yet result in what I see in the output! Lines 163-181 of keyword.asciidoc should appear for datasets, and apparently include "dataset-keyword-freetext.xml" twice and then dataset-keyword-controlled. dataset-freetext.xml says "satellite imagery" & "earth observation"; both outputs https://agiorguk.github.io/gemini/1063-gemini-services.html#6 & https://agiorguk.github.io/gemini/1062-gemini-datasets-and-data-series.html#6 show the same as each other - but quite different from the snippets (in the main branch) @archaeogeek : could you have a look at this last aspect?

**Comment by PeterParslow** [comment]
@archaeogeek is there a place to see the HTML output of the dev branch?

**Comment by archaeogeek** [comment]
> @archaeogeek is there a place to see the HTML output of the dev branch?

@PeterParslow the only way of doing this is to switch to the dev branch in the URL and then view the source of the page, eg https://agiorguk.github.io/gemini-dev/1062-gemini-datasets-and-data-series.html for the dev version of https://agiorguk.github.io/gemini/1062-gemini-datasets-and-data-series.html. Alternatively if you're able to build it locally then the html files are generated in your local folder

**Comment by archaeogeek** [comment]
@PeterParslow Looking into the issue with the incorrect keyword examples showing, in my local copy I have fixed the issue with the free text example showing twice, BUT there does seem to be a wider problem of the dataset variants being overwritten by the services variants, which is why you're seeing the same text in both cases. I honestly can't believe I would have missed this if it was a problem from the word go, so I'm trying to debug what's going on!

**Comment by archaeogeek** [comment]
I've figured out what's going on- and in the process a more efficient way of doing things. Rather than using the preprocessor approach we can use the tagged regions approach: https://docs.asciidoctor.org/asciidoc/latest/directives/include-tagged-regions/. This requires the following changes (for my benefit as much as anything else):

- [ ] Change all `ifdef::variant-dataset[]/endif::[]` references to `# tag::dataset[]/# end::dataset[]` in all the include files
- [ ] Change all `ifdef::variant-service[]/endif::[] `references to `# tag::service[]/# end::service[]` in all the include files
- [ ] In `1062-gemini-datasets-and-data-series` add a negative reference to the **service** tag in each include directive (this means it includes all regions EXCEPT the region tagged with service)
- [ ] In `1063-gemini-services` add a negative reference to the **dataset** tag in each include directive (this means it includes all regions EXCEPT the region tagged with dataset)
- [ ] Remove the double processing in the live github action so that it no longer processes everything twice- once for datasets and once for services
- [ ] Remove the `-a variant-dataset `parameter from the live github action
- [ ] Remove the double processing in the dev github action so that it no longer processes everything twice- once for datasets and once for services
- [ ] Remove the `-a variant-dataset` parameter from the dev github action

Consequently I will take this issue to work on these changes

**Comment by archaeogeek** [comment]
@PeterParslow can you review this and see if it still needs to be open please?

**Comment by PeterParslow** [comment]
Needs to stay open. There are currently no examples visible at https://agiorguk.github.io/gemini-dev/1063-gemini-services.html#6  and the issue still exists in main https://agiorguk.github.io/gemini/1063-gemini-services.html#6

This is true even if I 'view source' of the DEV HTML. so I'm not sure that the tag/include combination is working for this one.

**Comment by archaeogeek** [comment]
> Needs to stay open. There are currently no examples visible at https://agiorguk.github.io/gemini-dev/1063-gemini-services.html#6 and the issue still exists in main https://agiorguk.github.io/gemini/1063-gemini-services.html#6
> 
> This is true even if I 'view source' of the DEV HTML. so I'm not sure that the tag/include combination is working for this one.

Can you have another look please? I've just forced a re-run of the action

**Comment by PeterParslow** [comment]
Thanks. I can now see the examples in the DEV HTML. 
Please leave open & with me - I need to now fix the example to address this issue.

**Comment by PeterParslow** [comment]
See https://github.com/agiorguk/gemini/pull/140 of branch "Peter issue 29"

**Comment by PeterParslow** [comment]
Still to decide (in comment on July 29, 2021): whether to change "shall" to "should" in the encoding guidance, because the condition cannot be tested (whether the keyword is from a controlled vocabulary or not)

Implemented in DEV: changes to examples 2 & 3 for datasets, example 2 for services.
Still to pull into DEV: changes to examples 3 & 4 for services - I'd left that pull request as 'draft' (https://github.com/agiorguk/gemini/pull/135); changed that just today

**Comment by PeterParslow** [comment]
Issue remains in live example 4 - which claims to be an example of keywords from two controlled vocabularies, but does not cite the first one. Also, does not use the Anchor encoding for either.

**Comment by archaeogeek** [comment]
Actions:
1) General agreement on going from "shall" to "should"
2) @archaeogeek to change example 4 for services to include citation for humanCatalogueViewer

**Comment by archaeogeek** [comment]
Added fix for example 4 but- are we changing "shall" to "should" for both datasets and services, or just datasets?
